File: 5-longest-palindromic-substring/dynamic-programming/dynamic.c
#include <stdlib.h>
#include <string.h>  

char* longestPalindrome(char* s)
{
    int start       = 0;
    int maxLength   = 1;

    int low         = -1;
    int high        = -1;
    int N           = strlen(s);

    // Traverse
    for (int i = 0; i < N; i++)
    {
        // Process Odd Palindromes
        low = i;
        high = i;

        while (low >= 0 && high < N && s[low] == s[high])
        {
            int length = high - low + 1;

            if (length > maxLength)
            {
                start = low;
                maxLength = length;
            }

            low--;
            high++;
        }

        // Process Even Palindromes
        low = i;
        high = i + 1;

        while (low >= 0 && high < N && s[low] == s[high])
        {
            int length = high - low + 1;

            if (length > maxLength)
            {
                start = low;
                maxLength = length;
            }

            low--;
            high++;
        }
    }

    // Generate Substring
    char* sub = (char*)(malloc((maxLength + 1) * sizeof(char)));
    strncpy(sub, (s + start), maxLength);
    sub[maxLength] = '\0';
    return sub;
}

File: 5-longest-palindromic-substring/dynamic-programming/dynamic.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        start       = 0
        maxLength   = 1

        N           = len(s)
        arr = [[False] * N for x in range(N)]

        # Initialize Array
        for i in range(N):
            arr[i][i] = True
        
        # Check Matching Substrings of Length 2
        for i in range(N - 1):
            if s[i] == s[i + 1]:
                arr[i][i + 1] = True
                start = i
                maxLength = 2
        
        # Traverse
        for i in range(3, N + 1):
            for j in range(N - i + 1):
                k = i + j - 1

                if (arr[j + 1][k - 1]) and (s[j] == s[k]):
                    arr[j][k] = True

                    if i > maxLength:
                        start = j
                        maxLength = i
        
        return s[start:start + maxLength]

        